### PR TITLE
Testing with RSpec - Tic Tac Toe

### PR DESCRIPTION
**Pull request by shubham14p3**
We built test cases for the game logic in our Tic-Tac-Toe game, testing the public methods in the Board class mainly.

We focused on the Board class because it's the one responsible for all the game logic. i.e: Checking win and draw states, applying moves, checking valid and invalid moves, etc.

### CODE REVIEW

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Great Work @team :clap: However, some requirements need to be implemented in order for us to approve your project. Check :arrow_down: 

- **[REQUIRED]** The tests are not working, one reason is that you created a branch without using 'developement' branch as you base branch (i.e since it contains the approved code from the tic-tac-toe project without tests). the other is that you didn't provide instructions on how to set our local configuration to run them.
  - I suggest that you merge your 'development' branch with this branch and provide the necessary instructions to run the test locally (i.e Add them in the Readme file).

- **[REQUIRED]** Using `require_relative` for files in your directory instead of `require` is a ruby convention.
> "Require should be used for external files, like gems, while require_relative should be used for referring to files within your directory."

- **[REQUIRED]** Most of the commits were done by @shubham14p3 . The project is collaborative, would you explain to us why?.

### That's it :relieved: When finished, kindly resubmit another code review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

**Comment by shubham14p3** [comment]
@stratospherique  
1. As you mentioned about creating the branch without development branch which is not possible since "Development" branch is master itself now so every branch with be a sub tree.
2. [REQUIRED] Using require_relative for files in your directory instead of require is a ruby convention. but Require can also be used in Ruby , even I also got my another test case approved by using the same convention , it wont have any error.
3. [REQUIRED] Most of the commits were done by @shubham14p3 . The project is collaborative, would you explain to us why?. 
The reason is already told to Microverse Success Team #Greg , you can have an contact with them.
4. The reason the code was not running was because an player executable file was missing I have added the code , now it should work, Kindly check & let me know

If any major changes are required I will surely go forward and do the changes, but The feedback you suggested is not appropriate to make me any commit for the same

File: spec/board_rspec.rb [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+    it 'should return false if the move is out of boundaries' do
+      expect(my_board.valid_move?(10)).not_to eql(true)
+    end
+
+    it 'should return false if the position is already taken' do
+      my_board.apply_move?(my_player.character, 1)
+      expect(my_board.apply_move?(my_player.character, 1)).not_to eql(true)
+    end
+  end
```
Comment (stratospherique) [inline spec/board_rspec.rb]: * [ ] All the public methods of the classes implemented in project OOP are tested ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/Ruby_testing/all_tic_tac_toe.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59566098/))
  - 'valid_move?' is a private method, it breaks the tests when I run them. :pray: you can remove it if you can't turn it into public.

File: spec/board_rspec.rb [hunk 2]
```diff
@@ -0,0 +1,91 @@
+require './lib/board'
+require './bin/main'
```
Comment (stratospherique) [inline spec/board_rspec.rb]: - **[SUGGESTION]** - I was unable to run the tests until I removed the game initialization lines from main.rb file.  As a convention for this project, all classes should be kept in /lib folder. I suggest you add a file for your UI class in the /lib directory and import it from main.rb so you don't have to import the latter in your specs.

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Nice progress :clap: :clap: However, some changes need to be applied in order for us to approve your project. Check :arrow_down: 

- **[REQUIRED]** Check the code comments below, they indicate some issues that you need to address.

- [OPTIONAL] Since you implemented tests in your project, it's nice to add instructions on how to run them locally.

- **[OPTIONAL]** Using `require_relative` for files in your directory instead of `require` is a ruby convention.

> "Require should be used for external files, like gems, while require_relative should be used for referring to files within your directory."

### That's it :relieved: When finished, kindly resubmit another code review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

File: spec/board_rspec.rb (hunk 1 above)
Comment (shubham14p3) [reply spec/board_rspec.rb]: I gotcha what you mean, I have fixed the issues it was not accessing the files properly so the files were getting crashed, So It working fine now

File: spec/board_rspec.rb (hunk 2 above)
Comment (shubham14p3) [reply spec/board_rspec.rb]: Same as Above, file were missing added , you dont need to do any changes of removing game init so far.

**Review by mosesogwo — CHANGES_REQUESTED** [review]
**Hi @shubham14p3, I still cannot run your tests.**
- Please check the imports into your spec file. Some of the files required cannot load, the directory provided is not correct.
- Ensure that running `rspec spec/board_spec.rb`  from the terminal runs the tests successfully.
- Also, take note that the convention for naming spec files is `board_spec.rb`, not `board_rspec.rb`


_**Kindly make these changes and submit another code review request.**_

**Any question? You can reach me on Slack (Moses Ogwo)**

**Cheers:stars:**

### Status: Changes Requested.

**Comment by shubham14p3** [comment]
The problem for not getting executed was with the bin Main File which was converted into run time now I have converted it back to rb. Hopefully Now you would be able to run the commands with rspec also I have made the changes you asked for.

File: Gemfile [hunk 3]
```diff
@@ -0,0 +1,3 @@
+source 'https://rubygems.org'
+
+gem 'rubocop'
```
Comment (suretrust) [inline Gemfile:3]: `[Optional]` You should add `rspec` to your gemfile so it gets installed on running `bundle install`.

File: spec/board_spec.rb [hunk 4]
```diff
@@ -0,0 +1,91 @@
+require './lib/board'
+require './lib/player'
+require './bin/main'
```
Comment (suretrust) [inline spec/board_spec.rb:3]: `[Optional]` Use `require_relative` when importing within files.

**Review by suretrust — CHANGES_REQUESTED** [review]
Hi @shubham14p3,

### Great job thus far! 🙌 Please see my comments on your code. 👇 

- [ ] I couldn't get your tests to run, please ensure all tests run when you run `rspec` in the terminal or `rspec spec/board_spec.rb` when ran on a file. Please ensure to fix this before submitting for another code review as this seems like the same complaint other reviewers mentioned. See screenshot below:

Screenshot 2020-04-14 at 20 48 48


**Kindly submit for another code review once this has been fixed.** _Happy coding!_ 🎉

File: Gemfile (hunk 3 above)
Comment (shubham14p3) [reply Gemfile:3]: Added

**Comment by shubham14p3** [comment]
I have made the require changes so far I dont see any failures also, I was able to run the test cases as well & all test cases passed, kindly check on your end.

**Review by mosesogwo — APPROVED** [review]
**Congratulations :medal_sports:. Your project is approved!**

Please add instruction on how to run tests to your README before you merge.

### Status: Approved:heavy_check_mark:

**Pull request merged**